### PR TITLE
Error Categories.

### PR DESCRIPTION
**Pull request by 3Hren**
- [x] Out of scope shit SHOULD be eliminated.
- [x] Application-side support.

### CODE REVIEW

File: include/cocaine/api/connect.hpp [hunk 1]
```diff
@@ -80,12 +80,11 @@ class client:
...
     invoke(const std::shared_ptr<typename traits<Event>::dispatch_type>& dispatch, Args&&... args) {
         if(!m_session) {
-            throw cocaine::error_t("client is not connected");
+            throw std::system_error(std::make_error_code(std::errc::not_connected));
         }
 
-        if(std::is_same<typename result_of<Event>::type, io::mute_slot_tag>::value && dispatch) {
-            throw cocaine::error_t("callee has no upstreams specified");
-        }
+        BOOST_ASSERT((std::is_same<typename result_of<Event>::type, io::mute_slot_tag>::value)
```
Comment (3Hren) [inline include/cocaine/api/connect.hpp]: OOS.

File: include/cocaine/detail/unique_id.hpp [hunk 2]
```diff
@@ -30,9 +30,6 @@ namespace cocaine {
 struct unique_id_t {
     unique_id_t();
 
-    explicit
```
Comment (3Hren) [inline include/cocaine/detail/unique_id.hpp]: OOS.

File: include/cocaine/detail/unique_id.hpp [hunk 3]
```diff
@@ -43,10 +40,6 @@ struct unique_id_t {
     std::ostream&
     operator<<(std::ostream& stream, const unique_id_t& id);
 
-    static
```
Comment (3Hren) [inline include/cocaine/detail/unique_id.hpp]: OOS.

File: include/cocaine/rpc/dispatch.hpp [hunk 4]
```diff
@@ -245,8 +244,8 @@ dispatch<Tag>&
 dispatch<Tag>::on(const std::shared_ptr<io::basic_slot<Event>>& ptr) {
     typedef io::event_traits<Event> traits;
 
-    if(!m_slots->insert(std::make_pair(traits::id, ptr)).second) {
```
Comment (3Hren) [inline include/cocaine/rpc/dispatch.hpp]: OOS.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 5]
```diff
@@ -119,11 +119,19 @@ struct deferred {
...
 
     deferred&
-    abort(int code, const std::string& reason) {
-        outbox->synchronize()->template append<typename protocol::error>(code, reason);
+    abort(const std::error_code& ec, const std::string& reason) {
+        outbox->synchronize()->template append<typename protocol::error>(ec, reason);
         return *this;
     }
 
+#if defined(__clang__)
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: ?

File: include/cocaine/rpc/slot/deferred.hpp [hunk 6]
```diff
@@ -148,11 +156,19 @@ struct deferred<void> {
...
 
     deferred&
-    abort(int code, const std::string& reason) {
-        outbox->synchronize()->append<protocol::error>(code, reason);
+    abort(const std::error_code& ec, const std::string& reason) {
+        outbox->synchronize()->append<protocol::error>(ec, reason);
         return *this;
     }
 
+#if defined(__clang__)
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: ?

File: include/cocaine/rpc/slot/streamed.hpp [hunk 7]
```diff
@@ -49,11 +49,19 @@ struct streamed {
...
 
     streamed&
-    abort(int code, const std::string& reason) {
-        outbox->synchronize()->template append<typename protocol::error>(code, reason);
+    abort(const std::error_code& ec, const std::string& reason) {
+        outbox->synchronize()->template append<typename protocol::error>(ec, reason);
         return *this;
     }
 
+#if defined(__clang__)
```
Comment (3Hren) [inline include/cocaine/rpc/slot/streamed.hpp]: ?

File: include/cocaine/traits/dynamic.hpp [hunk 8]
```diff
@@ -34,7 +34,7 @@ namespace aux {
 
 template<class Stream>
 struct pack_dynamic:
-    public boost::static_visitor<>
+    public boost::static_visitor<void>
```
Comment (3Hren) [inline include/cocaine/traits/dynamic.hpp]: ?

File: include/cocaine/traits/dynamic.hpp [hunk 9]
```diff
@@ -42,7 +42,7 @@ struct pack_dynamic:
 
     void
     operator()(const dynamic_t::null_t& COCAINE_UNUSED_(source)) const {
-        m_target << msgpack::type::nil();
+        m_target.pack_nil();
```
Comment (3Hren) [inline include/cocaine/traits/dynamic.hpp]: OOS.

File: include/cocaine/traits/dynamic.hpp [hunk 10]
```diff
@@ -116,7 +116,7 @@ struct type_traits<dynamic_t> {
 
             for(; ptr < end; ++ptr) {
                 if(ptr->key.type != msgpack::type::RAW) {
-                    // NOTE: The keys should be strings.
```
Comment (3Hren) [inline include/cocaine/traits/dynamic.hpp]: OOS.

File: include/cocaine/traits/endpoint.hpp [hunk 11]
```diff
@@ -34,16 +34,16 @@ namespace cocaine { namespace io {
 template<class InternetProtocol>
 struct type_traits<asio::ip::basic_endpoint<InternetProtocol>> {
     typedef asio::ip::basic_endpoint<InternetProtocol> endpoint_type;
-    typedef std::tuple<std::string, unsigned short> tuple_type;
```
Comment (3Hren) [inline include/cocaine/traits/endpoint.hpp]: OOS.

File: include/cocaine/traits/endpoint.hpp [hunk 12]
```diff
@@ -34,16 +34,16 @@ namespace cocaine { namespace io {
...
+    typedef boost::mpl::list<std::string, unsigned short> sequence_type;
 
     template<class Stream>
     static inline
     void
     pack(msgpack::packer<Stream>& target, const endpoint_type& source) {
-        const std::string address = source.address().to_string();
-        const unsigned short port = source.port();
+        std::string address = source.address().to_string();
```
Comment (3Hren) [inline include/cocaine/traits/endpoint.hpp]: OOS.

File: include/cocaine/traits/endpoint.hpp [hunk 13]
```diff
@@ -34,16 +34,16 @@ namespace cocaine { namespace io {
...
+        unsigned short port = source.port();
 
-        type_traits<tuple_type>::pack(target, tuple_type(address, port));
```
Comment (3Hren) [inline include/cocaine/traits/endpoint.hpp]: OOS.

File: include/cocaine/traits/endpoint.hpp [hunk 14]
```diff
@@ -52,7 +52,7 @@ struct type_traits<asio::ip::basic_endpoint<InternetProtocol>> {
         std::string address;
         unsigned short port;
 
-        type_traits<tuple_type>::unpack(source, std::move(std::tie(address, port)));
```
Comment (3Hren) [inline include/cocaine/traits/endpoint.hpp]: OOS.

File: include/cocaine/traits/map.hpp [hunk 15]
```diff
@@ -48,15 +48,15 @@ struct type_traits<std::map<K, V>> {
             throw msgpack::type_error();
         }
 
+        std::pair<K, V> pair;
```
Comment (3Hren) [inline include/cocaine/traits/map.hpp]: OOS.

File: include/cocaine/traits/vector.hpp [hunk 16]
```diff
@@ -47,6 +47,7 @@ struct type_traits<std::vector<T>> {
             throw msgpack::type_error();
         }
 
+        // Replace the contents of the vector with the required number of copies of T.
```
Comment (3Hren) [inline include/cocaine/traits/vector.hpp]: OOS.

File: src/context.cpp [hunk 17]
```diff
@@ -250,34 +250,32 @@ context_t::bootstrap() {
...
             m_services->back().second->terminate();
             m_services->pop_back();
         }
 
-        COCAINE_LOG_ERROR(m_logger, "emergency core shutdown");
+        boost::spirit::karma::generate(builder, boost::spirit::karma::string % ", ", errored);
 
-        throw cocaine::error_t("couldn't start %d service(s)", errored.size());
-    } else {
-        m_signals.invoke<context::prepared>();
```
Comment (3Hren) [inline src/context.cpp]: Check here more carefully.

File: src/service/locator.cpp [hunk 18]
```diff
@@ -289,14 +289,12 @@ locator_t::locator_t(context_t& context, io_service& asio, const std::string& na
 
     // Service restrictions
 
-    if(!m_cfg.restricted.empty()) {
```
Comment (3Hren) [inline src/service/locator.cpp]: Check here.

File: src/service/locator.cpp [hunk 19]
```diff
@@ -317,13 +319,15 @@ locator_t::locator_t(context_t& context, io_service& asio, const std::string& na
...
+        const auto args = conf.at("args", dynamic_t::empty_object);
+
         m_cluster = m_context.get<api::cluster_t>(type, m_context, *this, name + ":cluster", args);
     }
 
     if(root.as_object().count("gateway")) {
         const auto conf = root.as_object().at("gateway").as_object();
         const auto type = conf.at("type", "unspecified").as_string();
-        const auto args = conf.at("args", dynamic_t::object_t());
+        const auto args = conf.at("args", dynamic_t::empty_object);
```
Comment (3Hren) [inline src/service/locator.cpp]: OOS, but this is the bug.

File: src/service/locator.cpp [hunk 20]
```diff
@@ -688,27 +690,25 @@ struct locator_category_t:
     virtual
     auto
     message(int code) const -> std::string {
-        switch(code) {
```
Comment (3Hren) [inline src/service/locator.cpp]: OOS.

File: src/service/node/app.cpp [hunk 21]
```diff
@@ -170,7 +170,7 @@ class app_service_t:
             typedef io::protocol<event_traits<app::enqueue>::upstream_type>::scope protocol;
             try {
                 upstream.send<protocol::error>(
-                    cocaine::error::dispatch_errors::service_error,
+                    cocaine::error::dispatch_errors::not_connected,
```
Comment (3Hren) [inline src/service/node/app.cpp]: Ensure.

File: src/session.cpp [hunk 22]
```diff
@@ -82,12 +82,16 @@ session_t::pull_action_t::finalize(const std::error_code& ec) {
     if(const auto ptr = *session->transport.synchronize()) {
 #endif
         try {
+            // NOTE: In case the underlying slot has miserably failed to handle its exceptions, the
```
Comment (3Hren) [inline src/session.cpp]: Seems like random code moving occurs here.

File: src/storage/files.cpp [hunk 23]
```diff
@@ -50,18 +50,22 @@ files_t::read(const std::string& collection, const std::string& key) {
...
     if(!fs::exists(file_path)) {
-        throw storage_error_t("object '%s' has not been found in '%s'", key, collection);
+        throw std::system_error(std::make_error_code(std::errc::no_such_file_or_directory),
+            file_path.string()
+        );
     }
 
     COCAINE_LOG_DEBUG(m_log, "reading object '%s'", key)(
         "collection", collection,
-        "path", file_path
```
Comment (3Hren) [inline src/storage/files.cpp]: OOS.

File: src/storage/files.cpp [hunk 24]
```diff
@@ -82,29 +86,29 @@ files_t::write(const std::string& collection, const std::string& key, const std:
     if(!fs::exists(store_status)) {
         COCAINE_LOG_INFO(m_log, "creating collection")(
             "collection", collection,
-            "path", store_path
+            "path",       store_path
```
Comment (3Hren) [inline src/storage/files.cpp]: OOS.

File: src/storage/files.cpp [hunk 25]
```diff
@@ -82,29 +86,29 @@ files_t::write(const std::string& collection, const std::string& key, const std:
...
+            store_path.string()
+        );
     }
 
     const fs::path file_path(store_path / key);
 
     COCAINE_LOG_DEBUG(m_log, "writing object '%s'", key)(
         "collection", collection,
-        "path", file_path
+        "path",       file_path
```
Comment (3Hren) [inline src/storage/files.cpp]: OOS.

File: src/storage/files.cpp [hunk 26]
```diff
@@ -148,14 +146,10 @@ files_t::remove(const std::string& collection, const std::string& key) {
 
     COCAINE_LOG_DEBUG(m_log, "removing object '%s'", key)(
         "collection", collection,
-        "path", file_path
+        "path",       file_path
```
Comment (3Hren) [inline src/storage/files.cpp]: Yet again OOS.

File: src/storage/files.cpp [hunk 27]
```diff
@@ -163,13 +157,13 @@ namespace {
 struct intersect {
     template<class T>
     T
-    operator()(const T& accumulator, T& keys) const {
+    operator()(const T& accu, T& keys) const {
```
Comment (3Hren) [inline src/storage/files.cpp]: WTF?

File: src/unique_id.cpp [hunk 28]
```diff
@@ -28,17 +28,6 @@ unique_id_t::unique_id_t() {
     uuid_generate(reinterpret_cast<unsigned char*>(uuid.data()));
 }
 
-unique_id_t::unique_id_t(const std::string& other) {
```
Comment (3Hren) [inline src/unique_id.cpp]: OOS.

File: src/unique_id.cpp [hunk 29]
```diff
@@ -58,12 +47,6 @@ unique_id_t::operator==(const unique_id_t& other) const {
            uuid[1] == other.uuid[1];
 }
 
-bool
```
Comment (3Hren) [inline src/unique_id.cpp]: Out of the scope.

**Comment by 3Hren** [comment]
Okay, now it looks better.

@antmat, take a look please.

File: include/cocaine/api/connect.hpp [hunk 30]
```diff
@@ -80,7 +80,7 @@ class client:
     typename traits<Event>::upstream_type
     invoke(const std::shared_ptr<typename traits<Event>::dispatch_type>& dispatch, Args&&... args) {
         if(!m_session) {
-            throw cocaine::error_t("client is not connected");
+            throw std::system_error(std::make_error_code(std::errc::not_connected));
```
Comment (antmat) [inline include/cocaine/api/connect.hpp]: Why do you omit messages here?

File: include/cocaine/api/storage.hpp [hunk 31]
```diff
@@ -97,13 +86,13 @@ storage_t::get(const std::string& collection, const std::string& key) {
     try {
         msgpack::unpack(&unpacked, blob.data(), blob.size());
     } catch(const msgpack::unpack_error& e) {
-        throw storage_error_t("object is corrupted");
+        throw std::system_error(std::make_error_code(std::errc::invalid_argument));
```
Comment (antmat) [inline include/cocaine/api/storage.hpp]: maybe create a helper like
make_system_error(std::errc::invalid_argument)
?

File: include/cocaine/detail/service/node/cached.hpp [hunk 32]
```diff
@@ -64,19 +64,19 @@ cached<T>::cached(context_t& context, const std::string& collection, const std::
 
     try {
         cache = api::storage(context, "cache");
-    } catch(const api::repository_error_t& e) {
+    } catch(const std::system_error& e) {
```
Comment (antmat) [inline include/cocaine/detail/service/node/cached.hpp]: Any checks for error code  here?

File: include/cocaine/errors.hpp [hunk 33]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+} // namespace error
+
+struct error_t:
+    public std::system_error
+{
+    template<class... Args>
+    error_t(const std::string& e, const Args&... args):
+        std::system_error(std::make_error_code(std::errc::invalid_argument), format(e, args...))
```
Comment (antmat) [inline include/cocaine/errors.hpp]: Why do you use invalid_argument here? Maybe unknown_category will fit better.

File: src/cluster/predefine.cpp [hunk 34]
```diff
@@ -57,7 +57,7 @@ struct dynamic_converter<predefine_cfg_t> {
         const dynamic_t& nodes = source.as_object().at("nodes", dynamic_t::object_t());
 
         if(nodes.as_object().empty()) {
-            throw cocaine::error_t("at least one node should be specified for predefined cluster");
+            throw cocaine::error_t("no nodes have been specified");
```
Comment (antmat) [inline src/cluster/predefine.cpp]: Why don't you use system_error here?

**Comment by antmat** [comment]
LGTM, except some minor notes

File: include/cocaine/api/connect.hpp (hunk 30 above)
Comment (3Hren) [inline include/cocaine/api/connect.hpp]: According to the http://en.cppreference.com/w/cpp/error/system_error/system_error a `what_arg` should contain an explanation string.

I think the `std::errc::not_connected` contains enough information to identify the error.

File: include/cocaine/api/storage.hpp (hunk 31 above)
Comment (3Hren) [inline include/cocaine/api/storage.hpp]: If the main reason is to save some code bytes, then I think that there is no need to indroduce another entity. If there are some other reasons, please explain.

Internally `std::system_error` has a constructor overload which accepts `std::error_code`, which can be constructed implicitly using error code enumeration if it implements `std::is_error_code` trait. However the `std::errc` implements only `std::is_error_condition` trait for some reasons.

File: include/cocaine/detail/service/node/cached.hpp (hunk 32 above)
Comment (3Hren) [inline include/cocaine/detail/service/node/cached.hpp]: For example?

Comment (antmat) [inline include/cocaine/detail/service/node/cached.hpp]: There was a specific catch for repository errors. Now we swallow any system_error, I mean maybe we should check that category is repository_error?

File: include/cocaine/errors.hpp (hunk 33 above)
Comment (3Hren) [inline include/cocaine/errors.hpp]: May be, but there is no other error code in a standard library that fits better. Personally I don't want to introduce another error code enum with a separate category just for this case.

File: src/cluster/predefine.cpp (hunk 34 above)
Comment (3Hren) [inline src/cluster/predefine.cpp]: The `cocaine::error_t` is already a system_error. There are ~50 throwing `error_t` in a codebase, mostly in constructors, so EINVAL fits perfectly for these cases.

File: include/cocaine/detail/service/node/cached.hpp (hunk 32 above)
Comment (3Hren) [inline include/cocaine/detail/service/node/cached.hpp]: Still don't understand :( Don't try to download if there are no "cache" storage?

**Comment by noxiouz** [comment]
It's impossible to comment lines outside a diff, so I do it here.
https://github.com/3Hren/cocaine-core/blob/nightly-ecat/src/service/locator.cpp#L693
This error also can mean that object which we get is corrupted and can't be unpacked by msgpack into a specified type. For example, a routing group with weights as floats.

**Comment by 3Hren** [comment]
Thanks! This one can be explained using optional string argument, I'll try to trigger that kind of error and to see how can we deal with it.

**Pull request closed without merging**